### DISCUSSION TITLE
Container dependency

### DISCUSSION THREAD

**Discussion by pionl** [JustSteveKing/php-sdk#12] [General]
Hi @JustSteveKing ,

This is what I was looking for, but have few questions about the possibility to make it more "framework" agnostic and more type strict.

1. I see you are using `phpfox/container` - would it be possible to remove the depedency and use symfony interface? __This way we could pass Laravel  / Symfony container.__

2. Also I see that you are binding $resource to container, it would be better to use `make` and provide the SDK as a paremeter to the "resolver". __This would allow me to use my own services in constructor.__

2. Instead of dynamic calls enfore more strict way how to add resources. -> I must implement method that would return the "resource" -> this would allow IDE to tell the user -> hey I've got these resources you can use (+ PHP Stan checks, etc)

What do you think?

Thanks

**Comment by JustSteveKing** [comment]
Hey @pionl

So I agree with all of those ideas!!

The phpfox/container was something I created specifically for this package! But I understand your points. I think a Container interface would work better here? 

I'd happily accept a PR or example on how you think some of this could work if you are interested?

**Reply by pionl** [reply]
Thank you for your quick input and confirmation - happy to have same opinion :)

I'm happy to make the PR. I will start tomorrow. I'll let you know.

Thanks for your time and the code 🚀

**Reply by pionl** [reply]
Need to study more the architecture if it fits, will let you know.

**Reply by JustSteveKing** [reply]
Let me know! I have a few ideas myself on how to improve this library - but I am not the only person to use it!

**Reply by pionl** [reply]
feel free to share.

Imho, improving the library is always good, you can release "new" version of it :) Those who prefer older syntax can use older version 👍

**Reply by pionl** [reply]
But I need to think about it, now I'm trying to understand the "path" system for the URI you have implemented :)

**Reply by pionl** [reply]
So I thought about the code and the direction. I would like to know, what is your direction.

As I understand it, your goal was to make non-generic way how a quick way to wrap the endpoints but leave it on the user to correctly find out how to build the "queries" on the sdk.

What I'm looking for is more strict way how to build an SDK, that would allow me to define exactly what are the inputs and what is the resource (can be generic but would be g8 to allow transformers to specified object).

Is this what you want?

For the container I would have ways how to improve it:

- at this moment you are requiring name in resource. This can cause problems because it should be unique but when sharing container there can be conflict. Easy option is to not require name, but just pass class name to container - it is unique
- we have to update `phpfox/container` dependency to v `^0.3.0` to allow usage in Laravel 9

**Reply by JustSteveKing** [reply]
So the main thing with the name is to allow calling the resource using `__call()` method, which is why we need a name, it feels more natural.

But I have been looking at ways to improve the developer experience of this, perhaps something like:

```php
$resource = SDK::make(Resource::class);

$response = $resource->get('/')->where('status', 'active')->send();
```

**Reply by pionl** [reply]
Looks nicer. Maybe `$sdk->make(...)` or use __invoke `$sdk(Resource::class)` ?

Still not decided what i prefer, sorry for that ;)

**Reply by pionl** [reply]
I hope I did not take to much of your time. I've decided to go with different aproach. If interested I can share the code when I'm done.

To fix the container issue I added "abstraction" around diff "containers". Something like this.

```php
class LaravelContainerFactory implements SDKContainerFactoryContract
{
    private Container $container;

    public function __construct(Container $container)
    {
        $this->container = $container;
    }

    public function makeEndpoint(AbstractApi $api, string $endpointClass): AbstractEndpoint
    {
        return $this->container->make($endpointClass, ['api' => $api]);
    }
}
```

**Reply by JustSteveKing** [reply]
I like the sound of this, it makes it a lot more flexible 😄

**Comment by pionl** [comment]
I've found that that container interface does not allow `make` with given 'sdk'. So unfortunately it is not possible.
